### DISCUSSION TITLE
Managed to get it working

### DISCUSSION THREAD

**Discussion by sigtel** [cariboulabs/cariboulite#156] [Q&A]
Apologies that I have not been over all the comments and this may be already well known but after a lot of struggle here are my notes having managed to get the thing working. Also, this is a memory dump while I still remember it all so probably a bit too verbose but there you are!

First off, I never succeeded on a Pi 3B. No matter what I tried I got nowhere - I never got the driver. So this is not particularly scientific as I redeployed a Pi 4...

Next, I never managed to get past the kernel issue when using Raspberry Pi OS on the Pi 4 (specifically the 64 bit Bookworm one)

Where I succeeded was with DragonOS on the Pi 4 (I even had issues with DragonOS on the Pi 3B - it never opened ssh for some reason but did on the Pi 4)

My method was this:
Install DragonOS and to an apt update / apt upgrade cycle (lots to download) and reboot.

Using https://github.com/konimaru/cariboulite get and install the software

This failed to complete with code errors.

Edit two source files: 
cariboulite/software/libcariboulite/src/CaribouLiteCpp.cpp 
and 
cariboulite/software/libcariboulite/src/CaribouLite.hpp 
- insert "#include <memory>" into each. This had caused the installation to bomb.

Strip out all occurrences of "apt" and "depmod" in install.sh (so it doesn't do them all again) and re-run install.sh. It ran through fine and gave me a test menu.

Edit /boot/firmware/config.txt and set these as this:
#dtparam=i2c_arm=on (commented out)
#dtparam=spi=on (commented out)
dtparam=i2c_vc=on (line inserted)
dtoverlay=spi1-3cs (line inserted)

Reboot.

I now had the driver - "lsmod|grep smi" showed this, and "/proc/device-tree/hat" now exists - I never had these before with any of my iterations.

And it works. Running "sudo SoapySDRUtils --find" succeeded, running "sudo ./cariboulite_test_app" all looks good (sorry I forgot where that is!), and, finally running "SoapySDRServer --bind" and running CubicSDR on my Mac and it finds the CaribouLite. Thus far I  have left it on a local radio station for several hours and it has not crashed.

Hope this helps someone.

**Comment by mvmn** [comment]
On Pi 3B with DragonOS: you can install from the develop_R1 branch (`git checkout develop_R1` before `install.sh`) - it installs without issues for me as long as you run it with sudo (otherwise it fails to access I2C on 7-th step or something like that).

Do necessary reboot - as a result the `/proc/device-tree/hat` appears.

But then I had the "couldn't open smi driver file '/dev/smi'" issue - which occurred if I do `sudo SoapySDRUtil --probe` or run `cariboulite_test_app` or anything like that.

I've just tried `/boot/firmware/config.txt` changes you did on Pi 3B and reboot - but it didn't help. Will also do apt update/upgrade.

P.S. BTW, lsmod | grep smi gives me this:
```
smi_stream_dev         24576 0
bcm2835_smi            24576 1 smi_stream_dev
```

P.P.S. So, for develop_R1 I keep getting "couldn't open smi driver file '/dev/smi'" issue.

**Comment by mvmn** [comment]
Tried with a fresh install of DragonOS Beta33 on Pi4 model B the main branch of Konimaru - still have the same issue.
I've modified `/boot/firmware/config.txt` as described, and also fixed install.sh to look for `/boot/firmware/config.txt` instead of `/boot/config.txt`

But I have a problem with kernel extension compilation, apparently:
```
3. SMI kernel module...
-- Kernel release: 6.5.0-1005-raspi
-- Kernel headers: /usr/src/linux-headers-6.5.0-1005-raspi
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ubuntu/projects/konimaru/cariboulite/software/libcariboulite/src/caribou_smi/kernel/build
compiling module smi_stream_dev.ko...
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: aarch64-linux-gnu-gcc-13 (Ubuntu 13.2.0-4ubuntu3) 13.2.0
  You are using:
/bin/sh: 1: gcc-13: not found
make[5]: *** [scripts/Makefile.build:251: /home/ubuntu/projects/konimaru/cariboulite/software/libcariboulite/src/caribou_smi/kernel/build/smi_stream_dev.o] Error 127
make[4]: *** [/usr/src/linux-headers-6.5.0-1005-raspi/Makefile:2037: /home/ubuntu/projects/konimaru/cariboulite/software/libcariboulite/src/caribou_smi/kernel/build] Error 2
make[3]: *** [Makefile:234: __sub-make] Error 2
make[2]: *** [CMakeFiles/smi_stream_dev.dir/build.make:81: smi_stream_dev] Error 2
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/smi_stream_dev.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

Also I tried adding sudo to cmake + make steps on "Main software" because it was writing errors about lack of access. It helped with no access issue - but instead I get SMI errors
```
12-03 18:09:33.224 30558 30558 E CARIBOU_SMI_MODULES caribou_smi_insert_smi_modules@caribou_smi_modules.c:71 Module insertion 'smi_stream_dev' failed
12-03 18:09:33.224 30558 30558 E CARIBOU_SMI caribou_smi_init@caribou_smi.c:456 Problem reloading SMI kernel modules
12-03 18:09:33.224 30558 30558 E CARIBOULITE Setup cariboulite_init_submodules@cariboulite_setup.c:280 Error setting up smi submodule
```

P.S. Damn - it's not finding gcc-13. And I can't install it - doesn't know that package. Let me do apt update/upgrade and retry...
P.P.S. Update/upgrade didn't help. I had to do this trick to install gcc-13
https://www.reddit.com/r/Ubuntu/comments/ptixle/is_there_a_ppa_that_always_has_the_latest_gcc/
```
sudo add-apt-repository ppa:ubuntu-toolchain-r/test
sudo apt-get update

version=$(apt-cache pkgnames | sed -nr 's/^gcc-(([0-9]+\.?)+)$/\1/p' \
                             | sort -n | tail -n1)
echo "The latest available GCC version: $version"

sudo apt-get install gcc-$version g++-$version
```

I wonder why does it even want gcc-13? What's wrong with gcc version 11?
In any case, the darn thing won't compile:
```
compiling module smi_stream_dev.ko...
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: aarch64-linux-gnu-gcc-13 (Ubuntu 13.2.0-4ubuntu3) 13.2.0
  You are using:           gcc-13 (Ubuntu 13.1.0-8ubuntu1~22.04) 13.1.0
In file included from ./include/linux/linkage.h:7,
                 from ./include/linux/kernel.h:17,
                 from /home/ubuntu/projects/cariboulite/software/libcariboulite/src/caribou_smi/kernel/build/smi_stream_dev.c:41:
/home/ubuntu/projects/cariboulite/software/libcariboulite/src/caribou_smi/kernel/build/smi_stream_dev.c: In function ‘smi_stream_dev_probe’:
./include/linux/export.h:29:22: error: passing argument 1 of ‘class_create’ from incompatible pointer type [-Werror=incompatible-pointer-types]
   29 | #define THIS_MODULE (&__this_module)
      |                     ~^~~~~~~~~~~~~~~
      |                      |
      |                      struct module *
/home/ubuntu/projects/cariboulite/software/libcariboulite/src/caribou_smi/kernel/build/smi_stream_dev.c:987:41: note: in expansion of macro ‘THIS_MODULE’
  987 |         smi_stream_class = class_create(THIS_MODULE, DEVICE_NAME);
      |                                         ^~~~~~~~~~~
In file included from ./include/linux/device.h:31,
                 from ./include/linux/platform_device.h:13,
                 from /home/ubuntu/projects/cariboulite/software/libcariboulite/src/caribou_smi/kernel/build/smi_stream_dev.c:44:
./include/linux/device/class.h:230:54: note: expected ‘const char *’ but argument is of type ‘struct module *’
  230 | struct class * __must_check class_create(const char *name);
      |                                          ~~~~~~~~~~~~^~~~
/home/ubuntu/projects/cariboulite/software/libcariboulite/src/caribou_smi/kernel/build/smi_stream_dev.c:987:28: error: too many arguments to function ‘class_create’
  987 |         smi_stream_class = class_create(THIS_MODULE, DEVICE_NAME);
      |                            ^~~~~~~~~~~~
./include/linux/device/class.h:230:29: note: declared here
  230 | struct class * __must_check class_create(const char *name);
      |                             ^~~~~~~~~~~~
cc1: some warnings being treated as errors
make[5]: *** [scripts/Makefile.build:251: /home/ubuntu/projects/cariboulite/software/libcariboulite/src/caribou_smi/kernel/build/smi_stream_dev.o] Error 1
make[4]: *** [/usr/src/linux-headers-6.5.0-1005-raspi/Makefile:2037: /home/ubuntu/projects/cariboulite/software/libcariboulite/src/caribou_smi/kernel/build] Error 2
make[3]: *** [Makefile:234: __sub-make] Error 2
make[2]: *** [CMakeFiles/smi_stream_dev.dir/build.make:81: smi_stream_dev] Error 2
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/smi_stream_dev.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

P.P.P.S. Fixed the Kernel module compilation error with these changes:
```
--- a/install.sh
+++ b/install.sh
@@ -107,15 +107,15 @@ printf "${CYAN}3. SMI kernel module...${NC}\n"
 cd $ROOT_DIR/software/libcariboulite/src/caribou_smi/kernel
 mkdir build
 cd build
-cmake ../
+CFLAGS=-Wno-error cmake ../
 make

 printf "${CYAN}3. Main software...${NC}\n"
 cd $ROOT_DIR
 mkdir build
 cd build
-cmake $ROOT_DIR/software/libcariboulite/
-make
+sudo cmake $ROOT_DIR/software/libcariboulite/
+sudo make
 sudo -u root make install

 # Configuration File
@@ -143,7 +143,7 @@ else
 fi

 printf "${GREEN}3. I2C-VC Configuration...  "
-DtparamSPI=`cat /boot/config.txt | grep "dtparam=i2c_vc" | xargs | cut -d\= -f1`
+DtparamSPI=`cat /boot/firmware/config.txt | grep "dtparam=i2c_vc" | xargs | cut -d\= -f1`
 if [ "$DtparamSPI" = "dtparam" ]; then
     printf "${CYAN}OK :)${NC}\n"
 else
diff --git a/software/libcariboulite/src/caribou_smi/kernel/smi_stream_dev.c b/software/libcariboulite/src/caribou_smi/kernel/smi_stream_dev.c
index 13281a6..9c26f3f 100644
--- a/software/libcariboulite/src/caribou_smi/kernel/smi_stream_dev.c
+++ b/software/libcariboulite/src/caribou_smi/kernel/smi_stream_dev.c
@@ -984,7 +984,7 @@ static int smi_stream_dev_probe(struct platform_device *pdev)
        }

        // Create sysfs entries with "smi-stream-dev"
-       smi_stream_class = class_create(THIS_MODULE, DEVICE_NAME);
+       smi_stream_class = class_create(DEVICE_NAME);
        ptr_err = smi_stream_class;
        if (IS_ERR(ptr_err))
        {
```
Also tried `./install.sh 2 2 3` instead of just `./install.sh`

However, the errors about module insertion remain:
```
12-03 18:57:09.392 64961 64961 E CARIBOU_SMI_MODULES caribou_smi_insert_smi_modules@caribou_smi_modules.c:71 Module insertion 'smi_stream_dev' failed
12-03 18:57:09.392 64961 64961 E CARIBOU_SMI caribou_smi_init@caribou_smi.c:434 Problem reloading SMI kernel modules
12-03 18:57:09.392 64961 64961 E CARIBOULITE Setup cariboulite_init_submodules@cariboulite_setup.c:280 Error setting up smi submodule
```

But, cariboulite_test_app works. At least I can toggle some LEDs and stuff - not sure how to check if it actually does something.
`sudo SoapySDRUtils --find` also returns these smi_stream_dev kernel module errors.

SDR++ though sees Caribou and seems to work (supposedly - I wasn't able to listen to any FM radio though so far, I only get noise) - but only if I run it with sudo. Otherwise it's not showing SoapySDR

**Reply by mvmn** [reply]
P.P.P.P.S. Seems like I can catch some stuff on below 1GHz receiver (I can see my PMR446 radio on a waterfall graph in SDR++), but nothing on the high frequency one (it actually shows a hump on whatever frequency I pick).

And even that below 1GHz fails to do stuff like decode FM radio.

**Comment by mvmn** [comment]
Ok, better way to do this on Pi4: Get DragonOS Beta 32 - not 33. This has older Kernel, which doesn't require manual fixes in the code to compile. But use the [Konimaru](https://github.com/konimaru/cariboulite) repo main branch - official CaribouLite develop_R1 compiles, but doesn't work (errors about SMI module and then timeouts).

Maybe I should also try main branch of the official repo, but for Konimaru on DragonOS beta 32 it compiles no issues - the only thing I modified was to remove root check from install.sh, correct path to config.txt to /boot/firmware/config.txt instead of just /boot/config.txt, and then I ran install.sh with sudo - and it all worked fine.

The capture software seems to work. The only problem for me is that I can't actually get any reception. On HiF receiver I always get a hump in the middle of the graph in SDR++ - despite chosen frequency. On Sub1GHz frequency I get pretty much nothing, not even that PRM 446 radio. Something is maybe wrong with the hardware itself - not sure.

P.S. Ok, I see that latest commit to the main branch (hash `b1ca9bf3ca7ff64b36dfc5d7c083961e3edfb096`) is 8 hours ago. I've tried the main branch `install.sh` without sudo on DragonOS Beta 32 now - and it seems to have installed OK - provided you have a clean DragonOS Beta 32 install. But I didn't have /dev/smi - so I tried install 2 2 3 there and stuff, got /dev/smi, but didn't get it to work in SDR software - I get errors etc.
